File: Leetcode/18.四数之和.cpp
/*
 * @lc app=leetcode.cn id=18 lang=cpp
 *
 * [18] 四数之和
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/*2021-10-01-21-20*/

//from https://github.com/fengwei2002/Algorithm
/* 
 !
 ! a b c d 互不相同的四元组
 *
 * 注意去重的写法
 */


// @lc code=start
class Solution {
   public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            if (i - 1 >= 0 && nums[i] == nums[i - 1]) continue;

            for (int j = i + 1; j < n; j++) {
                if (j - 1 >= i + 1 && nums[j] == nums[j - 1]) continue;

                int k = j + 1;
                int l = n - 1;

                while (k < l) {
                    if (nums[i] + nums[j] > target - (nums[k] + nums[l])) {
                        l--;
                    } else if (nums[i] + nums[j] + nums[k] + nums[l] < target) {
                        k++;
                    } else {
                        ans.push_back({nums[i], nums[j], nums[k], nums[l]});

                        while (k < l && nums[l] == nums[l - 1]) l--;
                        while (k < l && nums[k] == nums[k + 1]) k++;
                        l--, k++;
                    }
                }
            }
        }
        return ans;
    }
};
// @lc code=end

File: Leetcode/18.四数之和.go
/*
 * @lc app=leetcode.cn id=18 lang=golang
 *
 * [18] 四数之和
 */
package leetcode

import "sort"

/* --- 2022-04-07-10-02 --- */
// https://github.com/fengwei2002/Algorithm

// @lc code=start
func fourSum(nums []int, target int) [][]int {
	n := len(nums)
	ans := make([][]int, 0)
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})
	for i := 0; i < n; i++ {
		if i > 0 && nums[i] == nums[i - 1] {
			continue
		}
		for j := i + 1; j < n; j++ {
			if j > i + 1 && nums[j] == nums[j - 1] {
				continue
			}
			for k, l := j + 1, n - 1; k < l; k++ {
				if k > j + 1 && nums[k] == nums[k - 1] {
					continue
				}
				for k < l - 1 && nums[i] + nums[j] + nums[k] + nums[l - 1] >= target {
					l--
				}
				if nums[i] + nums[j] + nums[k] + nums[l] == target {
					ans = append(ans, []int{nums[i], nums[j], nums[k], nums[l]})
				}
			}
		}
	}
	return ans
}
// @lc code=end

